### PR TITLE
Added inventory plugin for NetBox

### PR DESCRIPTION
**Pull request by bdlamprecht**
This is an __*alpha*__ stage of using `NetBox` as a "Source-of-Truth" for a custom inventory.
My python skills are very limited, so please be nice :smile: 

In this example, I'm struggling to understand how to, if at all possible, name `groups` similar to the way the [tutorial](https://nornir.readthedocs.io/en/latest/tutorials/intro/inventory.html#Filtering-the-inventory) does it with `site`:

```
In [16]:    brg.filter(site="cmh").inventory.hosts.keys()
Out[16]:    dict_keys(['host1.cmh', 'host2.cmh', 'spine00.cmh', 'spine01.cmh', 'leaf00.cmh', 'leaf01.cmh'])
```

In this inventory, I can do the following examples (using the [Docker](https://github.com/ninech/netbox-docker) example of `NetBox`):
```
from nornir.core import Nornir
from netbox import NBInventory

nornir = Nornir(inventory=nb,dry_run='')

print("\nHosts:")
print(nornir.inventory.hosts)

print("\nGroups:")
print(nornir.inventory.groups)
```
And the output is as follows:
```
Hosts:
{'01d-vba-b': Host: 01d-vba-b, '00h-vba-a': Host: 00h-vba-a}

Groups:
{'wap': Group: wap, 'vba': Group: vba, 'via': Group: via, 'vsc': Group: vsc, 'vfa': Group: vfa, 'vid': Group: vid, 'ecd': Group: ecd, 'wlc': Group: wlc, 'vg': Group: vg, 'con': Group: con, 'vdd': Group: vdd, 'vda': Group: vda, 'vwd': Group: vwd, 'oob': Group: oob, 'vfw': Group: vfw, 'vpb': Group: vpb, 'end': Group: end, 'vtr': Group: vtr, 'srv': Group: srv, 'vsd': Group: vsd, 'eca': Group: eca, 'vod': Group: vod, 'vsa': Group: vsa, 'vua': Group: vua, 'vsl': Group: vsl}
```

My question is, can you create a custom inventory and name a group (such as `device_role`) to be able to filter it the same way that the tutorial does it for `site` and `role`?

### CODE REVIEW

**Comment by bdlamprecht** [comment]
Well, concerning your question about if the user would need to modify it before using, ideally no.  That is what I was trying to do with the `item.serialize()` on lines `22` and `41`.  I apologize for my lack of common knowledge concerning `python`, but that function basically returns all attributes that each `host` or `group` contains as `json` data.

The neat thing about NetBox, is every `device` (or `host`) **MUST** have a `device_role`, a `manufacturer`,  a `device_type` (model number) and be assigned to a `site`.

I don't have `python 3.6` installed anywhere yet, but I'll work on getting that done so I can use `black` to clear up any formatting issues.

As for this comment:
> You will be able to do that "soon".

Is there any timeline or _estimated_ date that that is expected to be implemented?

**Comment by dbarrosop** [comment]
> Is there any timeline or estimated date that that is expected to be implemented?

Hopefully it will happen during July.

I am willing to help out a bit with this, however, I am going to need you to do a few things first:

I need you to add a bash script under `nornir/tests/plugins/inventory/netbox/start.sh` that basically starts a docker container with netbox and populates some data via it's API using curl or similar. Once we have that we can integrate it with pytest so we can test your plugin as we do with nsot. I can do this last part myself, I just need you to help out giving me the shell script that can start the container and populates data.

**Comment by bdlamprecht** [comment]
I tried my best to create a Docker setup of `NetBox` and import sample test data as requested.
As for why it continues to fail on the check for `Black`, I'm not sure...
```
$ black nornir/tests/plugins/inventory/netbox/
All done! ✨ � ✨
13 files left unchanged.
```

Also, I don't know how to get `Travis` to test deployment of the `NetBox` container, but I've tested it, and as long as you have `pynetbox==3.4.1` installed (found in `tests/plugins/inventory/netbox/requirements.txt`), it should work and automatically import the data required.

Again, I'm not an experienced developer so please be nice about my code :smile:

**Comment by bdlamprecht** [comment]
@dbarrosop I've been making some progress on this `NBInventory` and I've been working on cleaning up all of the errors from `Travis-CI`.

Every file but one passes the `pylama` test and I'll fix that next week (except for the `netbox/configuration/configuration.py` which I've copied directly from `github.com/ninech/netbox`).

I don't want to alter that file, so I'm wondering, is there some way to tell the `pylama` check to ignore that file?

Any feedback would be greatly appreciated.

**Comment by dbarrosop** [comment]
I am going to try to create a `nornir-test-containers` repo where we can put the container generation there and upload them to docker hub, I don't want to generate them on the fly all the time or we will slow down everything.

File: nornir/plugins/inventory/netbox.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import os
+import pynetbox
```
Comment (dbarrosop) [inline nornir/plugins/inventory/netbox.py]: Don't use this, let's call directly the API using `requests`. As we are just making a few calls to the API and transforming the data to nornir format `pynetbox` doesn't really give us too much and introduces an extra dependency.

**Comment by dbarrosop** [comment]
Ok, this is way more involved than I initially thought. Let's change plans (sorry about that), let's start by replacing that custom python library with requests. Then we can use [requests-mock](https://pypi.org/project/requests-mock/) instead of the container. If you are not familiar with `requests-mock` I can help you with that, just let me know.

**Comment by bdlamprecht** [comment]
I _think_ I understand what you're doing here.

Are you uses a **static** DB dump here to use with a `docker` container:

However, as development of `netbox` (which uses `django` and `postgres` as a DB) progresses, I _believe_ there are going to be changes to the DB schema that are going to be important and it is going to be more work to keep a static DB dump than dynamically adding items via a `curl` call using `requests`.

a `static` DB similar to what is being done here:
https://github.com/nornir-automation/nornir/pull/172/files#diff-b67911656ef5d18c4ae36cb6741b7965L8
Also, since `netbox` uses `nginx` to proxy for both the WebUI and API, and again, for long-term maintence, I

**Comment by dbarrosop** [comment]
If you are unsure how to do that start by removing all the docker related code and migrate from `pynetbox` to `requests`. Then make a list of all the API calls your code is doing and create a file with the response content for each one of them and put them in a folder. Then we can take a look together.

**Comment by bdlamprecht** [comment]
Okay, sounds good.  I'll try to get to that starting tomorrow, thanks again.

**Comment by bdlamprecht** [comment]
As requested, I've removed `pynetbox` from the main inventory plugin (i.e. `nornir/plugins/inventory/netbox.py`).

However, I'm not sure exactly what the idea behind `requests-mock` is.
I spent a lot of time getting `docker` to work and don't want to waste my time just to have to re-do it again using `requests-mock`?

**Comment by dbarrosop** [comment]
> However, I'm not sure exactly what the idea behind `requests-mock`

Simplify the tests and make them faster

> I spent a lot of time getting docker to work and don't want to waste my time just to have to re-do it again using requests-mock?

Yeah, sorry about that, that's my fault. I didn't realize doing this with nsot was going to be so hard and add so much delay to the tests as many containers need to be downloaded :( I am going to merge this into a branch and do the fixes myself. Will open a PR and ping you once it's done.

**Pull request merged**